packages/runtime/src/widgets/PopupMenu.tsx

```diff
@@ -38,99 +51,145 @@ interface PopupMenuStyles {
+    const popupItems: MenuProps["items"] = [];
+
+    const items = values?.items;
+    if (items) {
+      const tempItems = items
+        .map((rawItem, index) => getMenuItem(rawItem, index))
+        .filter(Boolean);
 
       popupItems.push(...tempItems);
     }
 
-    if (isObject(itemTemplate) && !isEmpty(namedData)) {
-      const tempItems = namedData.map((item, index) => {
-        const value = evaluate<string | number>(
-          defaultScreenContext,
-          itemTemplate.value,
-          {
-            [itemTemplate.name]: get(item, itemTemplate.name) as unknown,
-          },
-        );
-
-        return (
-          <React.Fragment key={value}>
-            <Menu.Item key={value} onClick={() => action?.callback(item)}>
-              <CustomScopeProvider value={item as CustomScope}>
-                {EnsembleRuntime.render([itemTemplate.template])}
-              </CustomScopeProvider>
-            </Menu.Item>
-            {values?.showDivider
-              ? index < namedData.length - 1 && <Menu.Divider />
-              : null}
-          </React.Fragment>
-        );
+    if (isObject(itemTemplate) && !isEmpty(evaluatedNamedData.namedData)) {
+      const tempItems = evaluatedNamedData.namedData.map((item, index) => {
+        const itm: ItemType = {
+          key: `popupmenu_itemTemplate_${index}`,
+          label: (
+            <CustomScopeProvider value={item as CustomScope}>
+              {EnsembleRuntime.render([itemTemplate.template])}
+            </CustomScopeProvider>
+          ),
+        };
+        return itm;
       });
 
       popupItems.push(...tempItems);
     }
 
-    return (
-      <Menu className={values?.styles?.names} style={{ ...values?.styles }}>
-        {popupItems}
-      </Menu>
-    );
-  }, [values?.items, action, namedData, itemTemplate]);
+    if (values?.showDivider) {
+      for (let i = 1; i < popupItems.length; i += 2) {
+        popupItems.splice(i, 0, { type: "divider" });
+      }
+    }
+
+    return popupItems;
+  }, [
+    values?.items,
+    values?.showDivider,
+    itemTemplate,
+    evaluatedNamedData.namedData,
+    getMenuItem,
+  ]);
 
-  useEffect((): void => {
-    if (values && !values.widget) {
-      return;
+  const getWidget = useCallback(() => {
+    if (!values?.widget) {
+      throw Error("PopupMenu requires a widget to render the anchor.");
     }
-    // clone value so we're not updating the yaml doc
-    const widget = cloneDeep(values?.widget);
-    const actualWidget = unwrapWidget(widget!);
-    setWidgetProps(actualWidget);
-    // Only run once
-  }, []);
+    const widget = cloneDeep(values.widget);
+    const actualWidget = unwrapWidget(widget);
+    return EnsembleRuntime.render([actualWidget]);
+  }, [values?.widget]);
+
+  const handleMenuItemClick = useCallback<NonNullable<MenuProps["onClick"]>>(
+    ({ key }) => {
+      let item;
+      if (key.includes("itemTemplate")) {
+        const itemIndex = toNumber(key.split("_").at(-1));
```

Comment:
This will have issues for double digit items.

Reply:
how will this be an issue? 
~ already showing 20 items [here](https://github.com/EnsembleUI/ensemble-react/pull/796/files#diff-b34f6a6d35e67887babd0c3fd536a7232c76aa68f211264acee0db1a58057ddeR58)